### PR TITLE
[test-only][full-ci] ApiTest. creating a shared resource with an expiration date

### PR DESCRIPTION
**Pull request by ScharfViktor**
- [x] changed tests `shareSubItemOfSpace.feature` add ability to set expire date
- [x] refactoring code
- [x] deleted core test 
- [x] created test share to group
- [x] test for change expiration date
- [x] created bug https://github.com/owncloud/ocis/issues/5442
- [ ] test where user shares to group and to user with different role and expiration date. I haven't idea what should I expect in the `GET https://localhost:9200/ocs/v1.php/apps/files_sharing/api/v1/shares/shareId` response

### CODE REVIEW

**Comment by ownclouders** [comment]
:boom: Acceptance test [<strong>localApiTests-apiSpacesShares-ocis</strong>](https://drone.owncloud.com/owncloud/ocis/18347/46/1) failed. Further test are cancelled...

File: tests/acceptance/features/apiSpacesShares/copySpaces.feature [hunk 1]
```diff
@@ -731,7 +731,10 @@ Feature: copy file
     And user "Alice" has created a space "Project" with the default quota using the GraphApi
     And user "Alice" has created a folder "/newfolder" in space "Project"
     And user "Alice" has uploaded a file inside space "Project" with content "some content" to "/newfolder/personal.txt"
-    And user "Alice" shares the following entity "newfolder" inside of space "Project" with user "Brian" with role "editor"
+    And user "Alice" creates a share inside of space "Project" with settings:
+      | path      | newfolder |
+      | shareWith | Brian      |
+      | role      | editor   |
```
Comment (saw-jan) [inline tests/acceptance/features/apiSpacesShares/copySpaces.feature:737]:
```suggestion
      | path      | newfolder |
      | shareWith | Brian     |
      | role      | editor    |
```

File: tests/acceptance/features/apiSpacesShares/copySpaces.feature [hunk 2]
```diff
@@ -751,7 +754,10 @@ Feature: copy file
     And user "Alice" has created a folder "/newfolder" in space "Project"
     And user "Alice" has uploaded a file inside space "Project" with content "old content version 1" to "/newfolder/personal.txt"
     And user "Alice" has uploaded a file inside space "Project" with content "old content version 2" to "/newfolder/personal.txt"
-    And user "Alice" shares the following entity "newfolder" inside of space "Project" with user "Brian" with role "editor"
+    And user "Alice" creates a share inside of space "Project" with settings:
+      | path      | newfolder |
+      | shareWith | Brian      |
+      | role      | editor   |
```
Comment (saw-jan) [inline tests/acceptance/features/apiSpacesShares/copySpaces.feature:760]:
```suggestion
      | path      | newfolder |
      | shareWith | Brian     |
      | role      | editor    |
```

**Review by saw-jan — COMMENTED** [review]
Indentation only

File: tests/acceptance/features/bootstrap/Sharing.php [hunk 3]
```diff
@@ -2386,6 +2386,19 @@ public function userGetsAllTheSharesSharedWithHimUsingTheSharingApi(string $user
...
+	 * @When user :user gets share :share shared with him using the sharing API
+	 *
+	 * @param string $user
+	 *
+	 * @return void
+	 */
+	public function userGetsTheShareSharedWithHimUsingTheSharingApi(string $user, $share):void {
+		$user = $this->getActualUsername($user);
+    $shareId = $this->getLastPublicLinkShareId();
+    $this->getShareData($user, $shareId);
```
Comment (saw-jan) [inline tests/acceptance/features/bootstrap/Sharing.php:2399]:
```suggestion
	public function userGetsTheShareSharedWithHimUsingTheSharingApi(string $user, $share):void {
                $user = $this->getActualUsername($user);
                $shareId = $this->getLastPublicLinkShareId();
                $this->getShareData($user, $shareId);
```

File: tests/acceptance/features/bootstrap/Sharing.php [hunk 4]
```diff
@@ -2386,6 +2386,19 @@ public function userGetsAllTheSharesSharedWithHimUsingTheSharingApi(string $user
...
+  /**
+	 * @When user :user gets share :share shared with him using the sharing API
+	 *
+	 * @param string $user
+	 *
+	 * @return void
+	 */
+	public function userGetsTheShareSharedWithHimUsingTheSharingApi(string $user, $share):void {
+		$user = $this->getActualUsername($user);
+    $shareId = $this->getLastPublicLinkShareId();
```
Comment (saw-jan) [inline tests/acceptance/features/bootstrap/Sharing.php:2398]: Public link id for group share?

File: tests/acceptance/features/bootstrap/Sharing.php [hunk 5]
```diff
@@ -2386,6 +2386,19 @@ public function userGetsAllTheSharesSharedWithHimUsingTheSharingApi(string $user
...
+    $this->getShareData($user, $shareId);
+	}
```
Comment (individual-it) [inline tests/acceptance/features/bootstrap/Sharing.php:2400]: weird indention

File: tests/acceptance/features/apiSpacesShares/shareSubItemOfSpace.feature [hunk 6]
```diff
@@ -66,13 +83,35 @@ Feature: Share a file or folder that is inside a space
...
+      | role       | viewer                   |
+      | expireDate | 2042-01-01T23:59:59+0100 |
+    Then the HTTP status code should be "200"
+    And the OCS status code should be "200"
+    And the OCS status message should be "OK"
+    When user "Brian" accepts share "/folder" offered by user "Alice" using the sharing API
+    Then as "Brian" folder "Shares/folder" should exist
+    When user "Brian" gets share "Shares/folder" shared with him using the sharing API
+    Then the fields of the last response to user "Brian" should include
+      | expiration | 2042-01-01T23:59:59+0100 |
```
Comment (individual-it) [inline tests/acceptance/features/apiSpacesShares/shareSubItemOfSpace.feature:117]: What about combining both steps into one, to make it read easier and avoid When-Then-When-Then series?
```suggestion
    Then the information about the share "Shares/folder" for user "Brian" should include
      | expiration | 2042-01-01T23:59:59+0100 |
```

**Review by individual-it — DISMISSED** [review]

File: tests/acceptance/features/bootstrap/Sharing.php (hunk 4 above)
Comment (ScharfViktor) [reply tests/acceptance/features/bootstrap/Sharing.php:2398]: sorry, I didn't understand comment

File: tests/acceptance/features/bootstrap/Sharing.php (hunk 5 above)
Comment (ScharfViktor) [reply tests/acceptance/features/bootstrap/Sharing.php:2400]: I need help. I use visual studio code and I every time have problem with lint and indention: 
https://user-images.githubusercontent.com/84779829/214024185-c6738be4-e157-4eb6-aa1f-ee466e0b37be.mov

Comment (ScharfViktor) [reply tests/acceptance/features/bootstrap/Sharing.php:2400]: https://user-images.githubusercontent.com/84779829/214025592-d5517da4-322f-4017-abb1-45eea68868aa.mov

File: tests/acceptance/features/bootstrap/Sharing.php (hunk 4 above)
Comment (saw-jan) [reply tests/acceptance/features/bootstrap/Sharing.php]: Sorry for my concise comment :smile: 
What I meant to say is that you are using this step in this group share scenario `Scenario: A user shares the folder to the group`. It feels a bit confusing to see `getLastPublicLinkShareId()` being used to get the shareId of that group share.

Comment (ScharfViktor) [reply tests/acceptance/features/bootstrap/Sharing.php]: - admin shares resource to user or group(don't matter) and sets shareId as value `$this->setLastShareData()`
- recipient user (he can be group member) gets shareId `getLastPublicLinkShareId()` and sends request GET with shareId `https://localhost:9200/ocs/v1.php/apps/files_sharing/api/v1/shares/shareID` and checks response.

**Comment by ScharfViktor** [comment]
when creating the test, I noticed a difference in the expiration date between share and share public link
https://github.com/owncloud/ocis/blob/master/tests/acceptance/features/bootstrap/Sharing.php#L1292-L1296 

method `isFieldInResponse` cannot work same for share public link because expects in format: `2023-01-31 22:59:59`

actual in response
share: `<expiration>2023-01-31T23:59:59+0100</expiration>`
public link: ` <expiration>2023-01-31 22:59:59</expiration>`

in some body request `expireDate` is different so I changed test where I set the required time in the request and wait for the required time in the response

File: tests/acceptance/features/coreApiSharePublicLink3/allowGroupToCreatePublicLinks.feature [hunk 7]
```diff
@@ -49,18 +49,18 @@ Feature: public share sharers groups setting
     And parameter "public_share_sharers_groups_allowlist_enabled" of app "files_sharing" has been set to "yes"
     And parameter "public_share_sharers_groups_allowlist" of app "files_sharing" has been set to '["grp1"]'
     When user "Alice" updates the last public link share using the sharing API with
-      | expireDate | +3 days |
+      | expireDate | 2040-01-01T23:59:59+0100 |
```
Comment (saw-jan) [inline tests/acceptance/features/coreApiSharePublicLink3/allowGroupToCreatePublicLinks.feature:52]: Was this necessary?
Can we only assert the date (`2040-01-01`) but not the time in order to cope with the differences? That way we may use incremental days `+3`.

Comment (ScharfViktor) [reply tests/acceptance/features/coreApiSharePublicLink3/allowGroupToCreatePublicLinks.feature:52]: I did this for more transparency. If we use this aproach In the test, we can see what date format we use in the body of the request and what date format we get in the response. Unfortunately, we have a difference in this (I created a bug), and I have to postpone PR until it's clearer how it's supposed to work

Comment (ScharfViktor) [reply tests/acceptance/features/coreApiSharePublicLink3/allowGroupToCreatePublicLinks.feature:52]: we get a different date format in the response, so I:
- created error #5442
- check the date format (2040-01-01)

**Review by saw-jan — COMMENTED** [review]
LGTM

**Review by SagarGi — COMMENTED** [review]
lgtm :+1:

**Review by saw-jan — APPROVED** [review]

**Pull request merged**